analyzers/src/SonarAnalyzer.CSharp/Rules/MemberInitializerRedundant.RoslynCfg.cs

```diff
@@ -0,0 +1,111 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.CFG.Roslyn;
+using SonarAnalyzer.Extensions;
+using StyleCop.Analyzers.Lightup;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    public sealed partial class MemberInitializerRedundant
+    {
+        private class RoslynChecker : CfgAllPathValidator
+        {
+            private readonly ISymbol memberToCheck;
+            private readonly ControlFlowGraph cfg;
+
+            public RoslynChecker(ControlFlowGraph cfg, ISymbol memberToCheck) : base(cfg)
+            {
+                this.cfg = cfg;
+                this.memberToCheck = memberToCheck;
+            }
+
+            // Returns true if the block contains assignment before access
+            protected override bool IsValid(BasicBlock block) =>
+                Match(block, checkReadBeforeWrite: false);
+
+            // Returns true if the block contains access before assignment
+            protected override bool IsInvalid(BasicBlock block) =>
+                Match(block, checkReadBeforeWrite: true);
+
+            private bool Match(BasicBlock block, bool checkReadBeforeWrite)
+            {
+                foreach (var operation in block.OperationsAndBranchValue)
+                {
+                    if (ProcessOperation(operation, cfg, checkReadBeforeWrite, out var result))
+                    {
+                        return result;
+                    }
+                }
+
+                return false;
+            }
+
+            private bool ProcessOperation(IOperation operation, ControlFlowGraph controlFlowGraph, bool isRead, out bool result)
+            {
+                foreach (var child in operation.DescendantsAndSelf().ToReversedExecutionOrder())
+                {
+                    if (isRead && child.Instance.Kind == OperationKindEx.FlowAnonymousFunction)
+                    {
+                        var anonymousFunctionCfg = controlFlowGraph.GetAnonymousFunctionControlFlowGraph(IFlowAnonymousFunctionOperationWrapper.FromOperation(child.Instance));
+                        foreach (var subOperation in anonymousFunctionCfg.Blocks.SelectMany(x => x.OperationsAndBranchValue).SelectMany(x => x.DescendantsAndSelf()))
+                        {
+                            if (ProcessOperation(subOperation, anonymousFunctionCfg, isRead, out result))
+                            {
+                                return true;
+                            }
+                        }
```

Comment:
This is also similar to the comment on the other PR - https://github.com/SonarSource/sonar-dotnet/pull/4855#discussion_r709975109

>_Thinking out loud:_ This part is very similar to `anonymousFunctionCfg.Blocks.Any(FindReferenceToSelf)`, it just returns a little different result. 
>
>Can we unify them somehow? It's complicated so I'd like to have it just once there.

Reply:
The key idea is, that 
`anonymousFunctionCfg.Blocks.SelectMany(x => x.OperationsAndBranchValue).SelectMany(x => x.DescendantsAndSelf())` 
is basically the same as 
`foreach (var operation in block.OperationsAndBranchValue)` 
plus
 `foreach (var child in operation.DescendantsAndSelf().ToReversedExecutionOrder())` 
if you'd merge them based on the previous comment.

Reply:
You still need to fetch inner cfg to get its own operations, that's why this block is needed.  But indeed, I've simplified
anonymousFunctionCfg.Blocks.SelectMany(x => x.OperationsAndBranchValue).SelectMany(x => x.DescendantsAndSelf())
to
anonymousFunctionCfg.Blocks.SelectMany(x => x.OperationsAndBranchValue)
and the outer foreach does the trick we need

Reply:
I think you still don't see what I see in these two comments. After 3 or 4 rounds of decremental deduplications, this is what I see:

Delete line 42-88 (`IsValid` - updated, `IsInvalid` - updated, `Match` & `ProcessOperation` merged) and replace it with this:
```C#

            // Returns true if the block contains assignment before access
            protected override bool IsValid(BasicBlock block) =>
                ProcessBlock(block, cfg, false);

            // Returns true if the block contains access before assignment
            protected override bool IsInvalid(BasicBlock block) =>
                ProcessBlock(block, cfg, true);

            private bool ProcessBlock(BasicBlock block, ControlFlowGraph controlFlowGraph, bool checkReadBeforeWrite)
            {
                foreach (var operation in block.OperationsAndBranchValue.Reverse().ToReversedExecutionOrder())
                {
                    if (checkReadBeforeWrite && operation.Instance.Kind == OperationKindEx.FlowAnonymousFunction)
                    {
                        var anonymousFunctionCfg = controlFlowGraph.GetAnonymousFunctionControlFlowGraph(IFlowAnonymousFunctionOperationWrapper.FromOperation(operation.Instance));
                        if (anonymousFunctionCfg.Blocks.Any(x => ProcessBlock(x, anonymousFunctionCfg, checkReadBeforeWrite)))
                        {
                            return true;
                        }
                    }
                    else if (memberToCheck.Equals(MemberSymbol(operation.Instance)))
                    {
                        return IsReadOrWrite(operation, checkReadBeforeWrite);
                    }
                }
                return false;
            }
```